packages/gatsby/src/utils/worker/child/render-html.ts

```diff
@@ -413,3 +449,71 @@ export async function renderPartialHydrationProd({
     })
   }
 }
+
+export interface IRenderSliceResult {
+  chunks: 2 | 1
+}
+
+export interface IRenderSlicesResults {
+  [sliceName: string]: IRenderSliceResult
+}
+
+export interface ISlicePropsEntry {
+  sliceId: string
+  sliceName: string
+  props: Record<string, unknown>
+  hasChildren: boolean
+}
+
+export async function renderSlices({
+  slices,
+  htmlComponentRendererPath,
+  publicDir,
+  slicesProps,
+}: {
+  publicDir: string
+  slices: Array<[string, IGatsbySlice]>
+  slicesProps: Array<ISlicePropsEntry>
+  htmlComponentRendererPath: string
+}): Promise<void> {
+  const htmlComponentRenderer = require(htmlComponentRendererPath)
+
+  for (const { sliceId, props, sliceName, hasChildren } of slicesProps) {
+    const sliceEntry = slices.find(f => f[0] === sliceName)
+    if (!sliceEntry) {
+      throw new Error(
+        `Slice name "${sliceName}" not found when rendering slices`
+      )
+    }
+
+    const [_fileName, slice] = sliceEntry
+
+    const staticQueryContext = await readStaticQueryContext(
+      slice.componentChunkName
+    )
+
+    const MAGIC_CHILDREN_STRING = `__DO_NOT_USE_OR_ELSE__`
```

Comment:
So this refers to the React internal, this seems a bit brittle when they change things?